### PR TITLE
[MI-3300] Added API to unlink a channel

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**

### CODE REVIEW

File: server/getTeamsData.go [hunk 1]
```diff
@@ -217,6 +217,37 @@ func (p *Plugin) LinkChannels(userID, mattermostTeamID, mattermostChannelID, msT
...
+		p.API.LogError("Unable to get the current channel information.", "ChannelID", mattermostChannelID, "Error", appErr.Message)
+		return "Unable to get the current channel information.", http.StatusInternalServerError
+	}
+
+	if channel.Type == model.ChannelTypeDirect || channel.Type == model.ChannelTypeGroup {
+		p.API.LogError("Linking/unlinking a direct or group message is not allowed")
+		return "Linking/unlinking a direct or group message is not allowed", http.StatusBadRequest
+	}
+
+	canLinkChannel := p.API.HasPermissionToChannel(userID, mattermostChannelID, model.PermissionManageChannelRoles)
```
Comment (Kshitij-Katiyar) [inline server/getTeamsData.go:232]: same line

File: server/getTeamsData.go [hunk 2]
```diff
@@ -217,6 +217,37 @@ func (p *Plugin) LinkChannels(userID, mattermostTeamID, mattermostChannelID, msT
 	return "", http.StatusOK
 }
 
+func (p *Plugin) UnlinkChannels(userID, mattermostChannelID string) (string, int) {
+	channel, appErr := p.API.GetChannel(mattermostChannelID)
+	if appErr != nil {
+		p.API.LogError("Unable to get the current channel information.", "ChannelID", mattermostChannelID, "Error", appErr.Message)
+		return "Unable to get the current channel information.", http.StatusInternalServerError
```
[the remark this answers is not in the document]
Comment (raghavaggarwal2308) [reply server/getTeamsData.go:224]: discussed

File: server/api.go [hunk 3]
```diff
@@ -474,6 +478,24 @@ func (a *API) linkChannels(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func (a *API) unlinkChannels(w http.ResponseWriter, r *http.Request) {
+	userID := r.Header.Get(HeaderMattermostUserID)
+
+	pathParams := mux.Vars(r)
+	channelID := pathParams[PathParamChannelID]
```
Comment (ayusht2810) [inline server/api.go:485]: we should add validation for channelID here

File: server/api.go [hunk 4]
```diff
@@ -474,6 +478,24 @@ func (a *API) linkChannels(w http.ResponseWriter, r *http.Request) {
...
+
+	if errMsg, statusCode := a.p.UnlinkChannels(userID, channelID); errMsg != "" {
+		http.Error(w, errMsg, statusCode)
+		return
+	}
+
+	if _, err := w.Write([]byte("Channels unlinked successfully")); err != nil {
```
Comment (ayusht2810) [inline server/api.go:492]:
```suggestion
	if _, err := w.Write([]byte("Channel unlinked successfully")); err != nil {
```

File: server/api_test.go [hunk 5]
```diff
@@ -1268,3 +1268,74 @@ func TestLinkChannels(t *testing.T) {
...
+	for _, test := range []struct {
+		Name               string
+		SetupPlugin        func(*plugintest.API)
+		SetupStore         func(*storemocks.Store)
+		ExpectedResult     string
+		ExpectedStatusCode int
+		Body               string
+	}{
+		{
+			Name: "UnlinkChannels: channels linked successfully",
```
Comment (ayusht2810) [inline server/api_test.go:1282]:
```suggestion
			Name: "UnlinkChannels: channels unlinked successfully",
```

File: server/api_test.go [hunk 6]
```diff
@@ -1268,3 +1268,74 @@ func TestLinkChannels(t *testing.T) {
...
+			mockAPI := &plugintest.API{}
+			testutils.GetTeamsTeamID()
+			plugin.SetAPI(mockAPI)
+			defer mockAPI.AssertExpectations(t)
+
+			test.SetupPlugin(mockAPI)
+			test.SetupStore(plugin.store.(*storemocks.Store))
+
+			w := httptest.NewRecorder()
+			r := httptest.NewRequest(http.MethodDelete, fmt.Sprintf("/unlink-channels/%s", testutils.GetChannelID()), bytes.NewBufferString(test.Body))
```
Comment (ayusht2810) [inline server/api_test.go:1325]: why are we sending body here?
remove this from test cases as well

File: server/command_test.go [hunk 7]
```diff
@@ -100,6 +101,7 @@ func TestExecuteUnlinkCommand(t *testing.T) {
 			description: "Unable to get the current channel",
 			args:        &model.CommandArgs{},
 			setupAPI: func(api *plugintest.API) {
+				api.On("LogError", "Unable to get the current channel details.", "ChannelID", "", "Error", "").Times(1)
```
Comment (ayusht2810) [inline server/command_test.go:104]: why channel id and error are empty here?

**Review by Kshitij-Katiyar — APPROVED** [review]

File: server/command_test.go [hunk 8]
```diff
@@ -99,13 +99,17 @@ func TestExecuteUnlinkCommand(t *testing.T) {
...
+				UserId:    testutils.GetUserID(),
+				ChannelId: testutils.GetChannelID(),
+			},
 			setupAPI: func(api *plugintest.API) {
-				api.On("LogError", "Unable to get the current channel details.", "ChannelID", "", "Error", "").Times(1)
-				api.On("GetChannel", "").Return(nil, testutils.GetInternalServerAppError("Error while getting the current channel.")).Once()
-				api.On("SendEphemeralPost", "", &model.Post{
-					UserId:  "bot-user-id",
-					Message: "Unable to get the current channel details.",
+				api.On("LogError", "Unable to get the current channel details.", "ChannelID", testutils.GetChannelID(), "Error", "").Times(1)
```
Comment (ayusht2810) [inline server/command_test.go:107]: why error is empty here?

**Review by ayusht2810 — APPROVED** [review]

File: Makefile [hunk 9]
```diff
@@ -5,7 +5,7 @@ CURL ?= $(shell command -v curl 2> /dev/null)
 MM_DEBUG ?=
 MANIFEST_FILE ?= plugin.json
 GOPATH ?= $(shell go env GOPATH)
-GO_TEST_FLAGS ?= -race
```
Comment (manojmalik20) [inline Makefile:8]: Uncomment this.

File: server/api.go [hunk 10]
```diff
@@ -59,6 +62,7 @@ func NewAPI(p *Plugin, store store.Store) *API {
 	router.HandleFunc("/disconnect", api.handleAuthRequired(api.checkUserConnected(api.disconnect))).Methods(http.MethodGet)
 	router.HandleFunc("/linked-channels", api.handleAuthRequired(api.getLinkedChannels)).Methods(http.MethodGet)
 	router.HandleFunc("/link-channels", api.handleAuthRequired(api.checkUserConnected(api.linkChannels))).Methods(http.MethodPost)
+	router.HandleFunc(fmt.Sprintf("/unlink-channels/{%s}", PathParamChannelID), api.handleAuthRequired(api.checkUserConnected(api.unlinkChannels))).Methods(http.MethodDelete)
```
Comment (manojmalik20) [inline server/api.go:65]:
```suggestion
	router.HandleFunc("/channels/link", api.handleAuthRequired(api.checkUserConnected(api.linkChannels))).Methods(http.MethodPost)
	router.HandleFunc(fmt.Sprintf("/channels/{%s}/unlink", PathParamChannelID), api.handleAuthRequired(api.checkUserConnected(api.unlinkChannels))).Methods(http.MethodDelete)
```

File: server/get_teams_data.go [hunk 11]
```diff
@@ -230,6 +230,36 @@ func (p *Plugin) LinkChannels(userID, mattermostTeamID, mattermostChannelID, msT
...
+		p.API.LogError("Unable to unlink the channel, you have to be atleast a channel admin to unlink it.", "ChannelID", mattermostChannelID)
+		return "Unable to unlink the channel, you have to be atleast a channel admin to unlink it.", http.StatusForbidden
+	}
+
+	if _, err := p.store.GetLinkByChannelID(channel.Id); err != nil {
+		p.API.LogError("This Mattermost channel is not linked to any MS Teams channel.", "ChannelID", channel.Id, "Error", err.Error())
+		return "This Mattermost channel is not linked to any MS Teams channel.", http.StatusBadRequest
+	}
+
+	if err := p.store.DeleteLinkByChannelID(channel.Id); err != nil {
```
Comment (manojmalik20) [inline server/get_teams_data.go:255]: Please get this function in sync with the `main` branch.

File: server/api_test.go [hunk 12]
```diff
@@ -1330,3 +1330,71 @@ func TestLinkChannels(t *testing.T) {
...
+func TestUnlinkChannels(t *testing.T) {
+	for _, test := range []struct {
+		Name               string
+		SetupPlugin        func(*plugintest.API)
+		SetupStore         func(*storemocks.Store)
+		ExpectedResult     string
+		ExpectedStatusCode int
+	}{
+		{
+			Name: "UnlinkChannels: channels unlinked successfully",
```
Comment (manojmalik20) [inline server/api_test.go:1343]:
```suggestion
			Name: "TestUnlinkChannels: channels unlinked successfully",
```

File: server/api_test.go [hunk 13]
```diff
@@ -1330,3 +1330,71 @@ func TestLinkChannels(t *testing.T) {
...
+			SetupPlugin: func(api *plugintest.API) {
+				api.On("GetChannel", testutils.GetChannelID()).Return(&model.Channel{
```
Comment (manojmalik20) [inline server/api_test.go:1345]: I think there was a test util for getting the channel.

File: server/get_teams_data.go [hunk 14]
```diff
@@ -230,6 +230,36 @@ func (p *Plugin) LinkChannels(userID, mattermostTeamID, mattermostChannelID, msT
...
+	}
+
+	if channel.Type == model.ChannelTypeDirect || channel.Type == model.ChannelTypeGroup {
+		p.API.LogError("Linking/unlinking a direct or group message is not allowed")
+		return "Linking/unlinking a direct or group message is not allowed", http.StatusBadRequest
+	}
+
+	if !p.API.HasPermissionToChannel(userID, mattermostChannelID, model.PermissionManageChannelRoles) {
+		p.API.LogError("Unable to unlink the channel, you have to be atleast a channel admin to unlink it.", "ChannelID", mattermostChannelID)
+		return "Unable to unlink the channel, you have to be atleast a channel admin to unlink it.", http.StatusForbidden
```
Comment (manojmalik20) [inline server/get_teams_data.go:247]:
```suggestion
		return "Unable to unlink the channel, you have to be at least a channel admin to unlink it.", http.StatusForbidden
```

File: server/api.go (hunk 10 above)
Comment (manojmalik20) [inline server/api.go:65]: The user doesn't need to be connected for this.

File: server/get_teams_data.go (hunk 14 above)
Comment (raghavaggarwal2308) [reply server/get_teams_data.go:247]: What we have is fine!

File: server/api_test.go (hunk 12 above)
Comment (raghavaggarwal2308) [reply server/api_test.go:1343]: What we have is fine!

File: server/api.go [hunk 15]
```diff
@@ -188,7 +190,14 @@ func (a *API) autocompleteTeams(w http.ResponseWriter, r *http.Request) {
 	out := []model.AutocompleteListItem{}
 	userID := r.Header.Get(HeaderMattermostUserID)
 
-	teams, _, err := a.p.GetMSTeamsTeamList(userID, nil)
+	client, err := a.p.GetClientForUser(userID)
+	if err != nil {
+		data, _ := json.Marshal(out)
```
Comment (manojmalik20) [inline server/api.go:196]: Shouldn't we log the error here?

File: server/api.go [hunk 16]
```diff
@@ -220,8 +229,15 @@ func (a *API) autocompleteChannels(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	client, err := a.p.GetClientForUser(userID)
+	if err != nil {
+		data, _ := json.Marshal(out)
```
Comment (manojmalik20) [inline server/api.go:235]: Same.

File: server/api_test.go [hunk 17]
```diff
@@ -943,24 +943,24 @@ func TestGetLinkedChannels(t *testing.T) {
...
+				c.On("GetTeams", fmt.Sprintf("id in ('%s')", testutils.GetTeamsTeamID())).Return([]*msteams.Team{
 					{
-						ID:          testutils.GetTeamID(),
+						ID:          testutils.GetTeamsTeamID(),
 						DisplayName: "mock-name",
 					},
 				}, nil).Times(1)
 			},
-			ExpectedResult:     `[{"mattermostTeamID":"pqoeurndhroajdemq4nfmw","mattermostChannelID":"bnqnzipmnir4zkkj95ggba5pde","msTeamsTeamID":"pqoeurndhroajdemq4nfmw","msTeamsChannelID":"bnqnzipmnir4zkkj95ggba5pde","msTeamsTeamName":"mock-name","msTeamsChannelName":"mock-name"}]`,
+			ExpectedResult:     `[{"mattermostTeamID":"pqoeurndhroajdemq4nfmw","mattermostChannelID":"bnqnzipmnir4zkkj95ggba5pde","msTeamsTeamID":"testteam-team-team-team-testteamtest","msTeamsChannelID":"test-teams-channel","msTeamsTeamName":"mock-name","msTeamsChannelName":"mock-name"}]`,
```
Comment (manojmalik20) [inline server/api_test.go:959]: Please fix the Teams ID.

File: server/command.go [hunk 18]
```diff
@@ -150,7 +150,12 @@ func (p *Plugin) executeLinkCommand(args *model.CommandArgs, parameters []string
 		return p.cmdError(args.UserId, args.ChannelId, "Invalid link command, please pass the MS Teams team id and channel id as parameters.")
 	}
 
-	if errMsg, _ := p.LinkChannels(args.UserId, args.TeamId, args.ChannelId, parameters[0], parameters[1], nil); errMsg != "" {
+	client, err := p.GetClientForUser(args.UserId)
+	if err != nil {
+		return p.cmdError(args.UserId, args.ChannelId, "Unable to link the channel, looks like your account is not connected to MS Teams")
```
Comment (manojmalik20) [inline server/command.go:155]: Shouldn't we log the error here?

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**